tests/bsim/bluetooth/host/misc/hfc/src/main.c

```diff
@@ -0,0 +1,444 @@
+	 *
+	 * To do this, the DUT is connected to a peer that loops through sending
+	 * a few ATT notifications then disconnecting.
+	 *
+	 * The test stops after an arbitrary number of notifications have been
+	 * received.
+	 *
+	 * [verdict]
+	 * - no buffer allocation failures, timeouts or stalls.
+	 */
+	int err;
+	struct dut_state *s = &g_dut_state;
+
+	LOG_DBG("Test start: DUT");
+
+	s->rx = 0;
+
+	err = bt_enable(NULL);
+	ASSERT(err == 0, "Can't enable Bluetooth (err %d)\n", err);
+	LOG_DBG("Central: Bluetooth initialized.");
+
+	s->conn = connect_and_subscribe();
+
+	LOG_DBG("Central: Connected and subscribed to both peers");
+
+	/* Wait until we got all notifications from both peers */
+	while (s->rx < NUM_NOTIFICATIONS) {
+		LOG_DBG("%d packets left, waiting..",
+			NUM_NOTIFICATIONS - s->rx);
+		k_msleep(100);
+		if ((s->rx < NUM_NOTIFICATIONS) && is_disconnected(s->conn)) {
+			LOG_INF("reconnecting..");
+			/* release the ref we took in the `connected` callback */
+			bt_conn_unref(s->conn);
+
+			/* release the ref we took when starting the scanner */
+			bt_conn_unref(s->conn);
+
+			s->conn = NULL;
+			s->conn = connect_and_subscribe();
+		}
+	}
+
+	/* linux will "unref" the conn :p */
+	disconnect(s->conn);
+
+	PASS("DUT done\n");
+}
+
+void entrypoint_peer_0(void)
+{
+	int err;
+	int tx;
+	struct bt_conn *conn;
+	const struct bt_gatt_attr *attr;
+	uint8_t data[10];
+
+	LOG_DBG("Test start: peer 0");
+
+	err = bt_enable(NULL);
+	ASSERT(err == 0, "Can't enable Bluetooth (err %d)\n", err);
+	LOG_DBG("Bluetooth initialized.");
+
+	/* prepare data for notifications */
+	attr = &test_gatt_service.attrs[2];
+	memset(data, 0xfe, sizeof(data));
+
+	/* Pass unless something else errors out later */
+	PASS("peer 0 done\n");
+
+	tx = 0;
+	while (true) {
+		conn = connect_as_peripheral();
+
+		LOG_INF("wait until DUT subscribes");
+		UNSET_FLAG(is_subscribed);
+		WAIT_FOR_FLAG(is_subscribed);
+
+		LOG_INF("send notifications");
+		for (int ltx = 0; ltx < 10; ltx++) {
+			err = 1;
+			while (err) {
+				LOG_DBG("p0: TX %d", tx);
+				err = bt_gatt_notify(conn, attr, data, sizeof(data));
+			}
+			tx++;
+		}
+
+		k_msleep(50);
+
+		LOG_INF("disconnect");
+		err = disconnect(conn);
+		ASSERT(!err, "Failed to initate disconnect (err %d)", err);
+		bt_conn_unref(conn);
+		conn = NULL;
+	}
+}
+
+void test_tick(bs_time_t HW_device_time)
+{
+	bs_trace_debug_time(0, "Simulation ends now.\n");
+	if (bst_result != Passed) {
+		bst_result = Failed;
+		bs_trace_error("Test did not pass before simulation ended.\n");
+	}
+}
+
+void test_init(void)
+{
+	bst_ticker_set_next_tick_absolute(TEST_TIMEOUT_SIMULATED);
+	bst_result = In_progress;
+}
+
+static const struct bst_test_instance test_to_add[] = {
+	{
+		.test_id = "dut",
+		.test_pre_init_f = test_init,
+		.test_tick_f = test_tick,
+		.test_main_f = entrypoint_dut,
+	},
+	{
+		.test_id = "peer_0",
+		.test_pre_init_f = test_init,
+		.test_tick_f = test_tick,
+		.test_main_f = entrypoint_peer_0,
```

Comment:
In the absence of `peer_1`, shouldn't this just be `peer`? 

Would it make sense to rename them central and peripheral, as the `dut` clearly is the central and the `peer` is the peripheral?

Reply:
I still prefer to have "dut". It's always hard to know what is being tested in the "central" and "peripheral" tests. Eg. gatt tests.

Reply:
> It's always hard to know what is being tested in the "central" and "peripheral" tests. Eg. gatt tests.

That is kind of true, but given that it's the sample implement on both sides, both are technically `dut` as well :D We do have the `bst_test_instance.test_descr` for the purpose of describing the tests, which we aren't using that much